lib/dialogue-agent/user-input.ts

```diff
@@ -56,26 +57,29 @@ function parseSpecial(intent : Ast.SpecialControlIntent, context : { platformDat
  * a UI action (a button) or a natural language command.
  */
 class UserInput {
+    utterance : string|null;
     platformData : PlatformData;
 
-    constructor(platformData : PlatformData) {
+    constructor(utterance : string|null, platformData : PlatformData) {
+        this.utterance = utterance;
         this.platformData = platformData;
     }
 
-    static fromThingTalk(thingtalk : Ast.Input, context : { platformData : PlatformData }) : UserInput {
+    static fromThingTalk(thingtalk : Ast.Input,
+                         context : { command : string|null, platformData : PlatformData }) : UserInput {
         if (thingtalk instanceof Ast.ControlCommand) {
             if (thingtalk.intent instanceof Ast.SpecialControlIntent)
                 return parseSpecial(thingtalk.intent, context);
             else if (thingtalk.intent instanceof Ast.AnswerControlIntent)
-                return new UserInput.Answer(thingtalk.intent.value, context.platformData);
+                return new UserInput.Answer(thingtalk.intent.value, context.command, context.platformData);
             else if (thingtalk.intent instanceof Ast.ChoiceControlIntent)
-                return new UserInput.MultipleChoiceAnswer(thingtalk.intent.value, context.platformData);
+                return new UserInput.MultipleChoiceAnswer(thingtalk.intent.value, context.command, context.platformData);
             else
                 throw new TypeError(`Unrecognized bookkeeping intent`);
         } else if (thingtalk instanceof Ast.Program) {
-            return new UserInput.Program(thingtalk, context.platformData);
+            return new UserInput.Program(thingtalk, context.command!, context.platformData);
```

Comment:
Yeah, all these null assertions are iffy. Either things can be null, or they can't.

Reply:
> Yeah, all these null assertions are iffy. Either things can be null, or they can't.

Different subclasses of `UserInput` have different types. E.g., `Answer` can be generated by UI pickers, so it could be null; while Program can only be generated by either neural network or with `\t` (at least currently) - so there's always an "utterance". Do you think we should just relax it and allow all subclasses to have null utterance?

Reply:
I think so. It's very hard to prove that the code is correct otherwise, and if the code is not correct, you're making the type system unsound which is even worse than no type system at all.